File: apps/driver-android/src/main/kotlin/lk/mageride/driver/tracker/TrackerPairingScreen.kt
package lk.mageride.driver.tracker

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.outlined.ArrowBack
import androidx.compose.material.icons.outlined.QrCodeScanner
import androidx.compose.material.icons.outlined.SatelliteAlt
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuAnchorType
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.KeyboardType
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.ui.component.DashboardBanner
import lk.mageride.driver.ui.component.LabelledTextField
import lk.mageride.driver.ui.component.MageRideCta
import lk.mageride.driver.ui.component.NoticeCard
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.driver.ui.vehicleLabel
import lk.mageride.shared.data.models.registry.VehicleSummary
import org.koin.androidx.compose.koinViewModel

/**
 * **SCR-DA-027 · GPS tracker pairing** (US-3.1/3.2, US-3.21–3.23, T-02/T-09).
 *
 * The wireframe, top to bottom: a `‹ Pair GPS tracker` app bar, the **Vehicle** selector, the
 * **IMEI** field, a two-button row (`▣ Scan device QR` · `Bind code`), the `Pair device` CTA, the
 * green *"✦ Hardware tracker behaviour"* card, and the fleet-CSV note.
 *
 * **Two of the wireframe's controls do less than it draws, and both are the wrapper's doing.**
 * `POST /v1/vehicles/{vehicleId}/device` — the only tracker route this app has — takes `{ imei }`
 * and nothing else, while provisioning-svc's own bind takes `method: [manual, qr, admin_code]` and
 * a `bindCode`. So **Bind code** has nothing to send and is drawn disabled with the reason under
 * it, and a scanned IMEI reaches the server indistinguishable from a typed one. Both are recorded
 * as C074 spec gaps; see [TrackerRepository].
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun TrackerPairingScreen(onBack: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: TrackerPairingViewModel = koinViewModel()
    val state by viewModel.state.collectAsStateWithLifecycle()

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(text = stringResource(R.string.tracker_title)) },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Outlined.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
            )
        },
    ) { insets ->
        Column(
            modifier = Modifier
                .padding(insets)
                .fillMaxSize()
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            // US-3.4 / T-08. A duplicate IMEI is not a failure to retry — the serial is already
            // active somewhere and provisioning-svc has quarantined it — so it gets its own copy
            // rather than the generic "try again".
            if (state.quarantined) {
                DashboardBanner(
                    text = stringResource(R.string.tracker_quarantined),
                    accent = MaterialTheme.colorScheme.error,
                )
            } else {
                state.error?.let { message ->
                    DashboardBanner(text = stringResource(message), accent = MaterialTheme.colorScheme.error)
                }
            }

            Column(
                modifier = Modifier.padding(MageRideTheme.spacing.md),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.sm),
            ) {
                VehicleSelector(
                    vehicles = state.vehicles,
                    selected = state.selected,
                    onSelect = viewModel::selectVehicle,
                )

                state.binding?.let { binding -> PairedConfirmation(binding = binding) }

                LabelledTextField(
                    label = stringResource(R.string.tracker_imei_label),
                    value = state.imei,
                    onValueChange = viewModel::onImeiChange,
                    isError = state.imeiRejected,
                    supporting = stringResource(
                        if (state.imeiRejected) R.string.tracker_imei_invalid else R.string.tracker_imei_help,
                    ),
                    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
                )

                EntryMethodRow(enabled = !state.isPaired, onScan = viewModel::startScan)

                MageRideCta(
                    label = stringResource(R.string.tracker_pair_action),
                    onClick = viewModel::pair,
                    enabled = state.canPair,
                    loading = state.pairing,
                )

                HardwareBehaviourCard()
                FleetCsvCard()
            }
        }
    }

    if (state.scanning) {
        DeviceQrScannerDialog(onScanned = viewModel::onScanned, onDismiss = viewModel::cancelScan)
    }
}

/**
 * The wireframe's `Vehicle: [Three-wheeler ▾]`.
 *
 * Every vehicle the driver owns or has been assigned is offered, Mode A/B included: a tracker on a
 * fleet bus is precisely the US-3.22 case, and the app's Mode-C-only fence (AL-27) is about
 * *onboarding* a vehicle, not about pairing hardware to one somebody else onboarded.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun VehicleSelector(
    vehicles: List<VehicleSummary>,
    selected: VehicleSummary?,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
) {
    var expanded by remember { mutableStateOf(false) }

    ExposedDropdownMenuBox(
        expanded = expanded,
        onExpandedChange = { expanded = it },
        modifier = modifier,
    ) {
        OutlinedTextField(
            value = selected?.let { vehicleLabel(it) }.orEmpty(),
            onValueChange = {},
            modifier = Modifier
                .fillMaxWidth()
                // `PrimaryNotEditable`: the field is the menu's anchor and is read-only, so
                // TalkBack announces a dropdown rather than a text box that refuses typing. The
                // no-argument `menuAnchor()` is deprecated in M3 1.4.
                .menuAnchor(ExposedDropdownMenuAnchorType.PrimaryNotEditable),
            readOnly = true,
            label = { Text(text = stringResource(R.string.tracker_vehicle_label)) },
            supportingText = if (vehicles.isEmpty()) {
                { Text(text = stringResource(R.string.tracker_no_vehicles)) }
            } else {
                null
            },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            shape = RoundedCornerShape(MageRideTheme.radius.md),
        )

        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            vehicles.forEach { vehicle ->
                DropdownMenuItem(
                    text = { Text(text = vehicleLabel(vehicle)) },
                    onClick = {
                        onSelect(vehicle.vehicleId)
                        expanded = false
                    },
                )
            }
        }
    }
}

/**
 * D2' §SCR-DA-027's *"paired → cert-issued confirmation"*.
 *
 * The `201` from the bind **is** the credential: provisioning-svc mints the X.509 or the signed PSK
 * inside that call (T-02, D6' §4.2), so a `bindingId` in hand means one was issued. What this card
 * cannot show is the device's live health — `lastSeen`, `battery`, `signal` come from
 * `GET /v1/trackers/{imei}`, which has no client here (C074 spec gap).
 */
@Composable
private fun PairedConfirmation(binding: TrackerBinding, modifier: Modifier = Modifier) {
    NoticeCard(
        accent = MageRideTheme.status.success,
        modifier = modifier,
        icon = Icons.Outlined.SatelliteAlt,
        title = stringResource(R.string.tracker_paired_title),
    ) {
        Text(
            text = stringResource(R.string.tracker_paired_body, TrackerImei.grouped(binding.imei)),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/** The wireframe's `[▣ Scan device QR] [Bind code]` row. */
@Composable
private fun EntryMethodRow(enabled: Boolean, onScan: () -> Unit, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
            OutlinedButton(onClick = onScan, enabled = enabled, modifier = Modifier.weight(1f)) {
                Icon(
                    imageVector = Icons.Outlined.QrCodeScanner,
                    contentDescription = null,
                    modifier = Modifier.padding(end = MageRideTheme.spacing.xxs),
                )
                Text(text = stringResource(R.string.tracker_scan_action))
            }
            // Disabled, and the line underneath says why: the owner-facing wrapper carries no
            // `bindCode`, so there is nowhere for one to be sent.
            OutlinedButton(onClick = {}, enabled = false, modifier = Modifier.weight(1f)) {
                Text(text = stringResource(R.string.tracker_bind_code_action))
            }
        }
        Text(
            text = stringResource(R.string.tracker_bind_code_help),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

/**
 * The wireframe's green *"✦ Hardware tracker behaviour"* card — and the C074 fence, in words.
 *
 * US-3.6: once the device is bound it is the single active publisher and the phone stops.
 * US-3.22/3.23: a Mode A/B tracker starts and ends journeys on ignition with no app involved, so
 * the dashboard is *already* at "Journey started" when the driver opens it. US-3.21: a Mode C
 * tracker's GPS is accepted only while the driver is Online.
 */
@Composable
private fun HardwareBehaviourCard(modifier: Modifier = Modifier) {
    NoticeCard(
        accent = MageRideTheme.status.success,
        modifier = modifier,
        icon = Icons.Outlined.SatelliteAlt,
        title = stringResource(R.string.tracker_behaviour_title),
    ) {
        Column(verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs)) {
            listOf(
                R.string.tracker_behaviour_publisher,
                R.string.tracker_behaviour_ignition,
                R.string.tracker_behaviour_mode_c,
            ).forEach { line ->
                Text(
                    text = stringResource(line),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

/**
 * The wireframe's *"Fleet (5,000+ trackers)? Use the Admin Portal CSV upload ›"* (US-3.2, T-09).
 *
 * Drawn as a note rather than as a link: the bulk path is `POST /v1/fleets/{fleetId}/trackers/bulk`
 * on provisioning-svc and it is the **Admin Portal's**, and this app has no Admin Portal origin to
 * open — `DriverEnvironment` carries the gateway and the MQTT host, and D7' puts the portal on a
 * different one. A `›` that opened nothing would be worse than a sentence that says where to go.
 */
@Composable
private fun FleetCsvCard(modifier: Modifier = Modifier) {
    NoticeCard(accent = MaterialTheme.colorScheme.secondary, modifier = modifier) {
        Text(
            text = stringResource(R.string.tracker_fleet_csv),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

File: apps/driver-ios/DriverApp/Tracker/TrackerPairingScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-027 · GPS tracker pairing** (US-3.1/3.2, US-3.21–3.23, T-02/T-09).
///
/// The wireframe, top to bottom: a `‹ Back` / *"Pair GPS tracker"* bar, one grouped list holding
/// **Vehicle** and **IMEI**, a two-button row (`▣ Scan device QR` · `Bind code`), the `Pair device`
/// CTA, the green *"✦ Hardware tracker behaviour"* card, and the fleet-CSV note.
///
/// **Two of the wireframe's controls do less than it draws, and both are the wrapper's doing.**
/// `POST /v1/vehicles/{vehicleId}/device` — the only tracker route this app has — takes `{ imei }` and
/// nothing else, while provisioning-svc's own bind takes `method: [manual, qr, admin_code]` and a
/// `bindCode`. So **Bind code** has nothing to send and is drawn disabled with the reason under it,
/// and a scanned IMEI reaches the server indistinguishable from a typed one. Both are C074 spec gaps
/// carried forward; see ``TrackerRepository``.
///
/// `@MainActor` on the whole view, not on its initialiser — see ``ProfileSetupScreen`` for why.
@MainActor
struct TrackerPairingScreen: View {

    @StateObject private var model: TrackerPairingModel

    init(model: @autoclosure @escaping () -> TrackerPairingModel) {
        _model = StateObject(wrappedValue: model())
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: MageRideSpacing.sm) {
                banner

                GroupedList {
                    vehicleRow
                    imeiRow
                }

                if let binding = model.state.binding {
                    pairedCard(binding)
                }

                entryMethods

                Button {
                    Task { await model.pair() }
                } label: {
                    Text(key: "tracker_pair_action")
                }
                .buttonStyle(.mageCta(loading: model.state.isPairing))
                .disabled(!model.state.canPair)

                hardwareBehaviour
                fleetCsv
            }
            .padding(MageRideSpacing.md)
        }
        .frame(maxWidth: .infinity)
        .background(MageRideColor.background)
        .navigationTitle(Text(key: "tracker_title"))
        .navigationBarTitleDisplayMode(.inline)
        .task { await model.refresh() }
        .refreshable { await model.refresh() }
        .sheet(
            isPresented: Binding(
                get: { model.state.isScanning },
                set: { if !$0 { model.cancelScan() } }
            )
        ) {
            DeviceQrScannerSheet(onScanned: model.onScanned, onDismiss: model.cancelScan)
        }
    }

    // MARK: - The strip

    /// US-3.4 / T-08. A duplicate IMEI is not a failure to retry — the serial is already active
    /// somewhere and provisioning-svc has quarantined it — so it gets its own copy rather than the
    /// generic *"try again"*. Everything else lands as the resolved failure copy (D-26).
    @ViewBuilder
    private var banner: some View {
        if let errorKey = model.state.errorKey {
            DashboardBanner(
                text: errorKey.localised,
                accent: MageRideColor.error,
                symbolName: model.state.isQuarantined ? "exclamationmark.shield.fill" : nil
            )
            .onTapGesture(perform: model.dismissError)
        }
    }

    // MARK: - The two rows

    /// The wireframe's `Vehicle   ABC-1234 ›`.
    ///
    /// Every vehicle the driver owns or has been assigned is offered, Mode A/B included: a tracker on a
    /// fleet bus is precisely the US-3.22 case, and the app's Mode-C-only fence (AL-27) is about
    /// *onboarding* a vehicle, not about pairing hardware to one somebody else onboarded.
    ///
    /// A menu `Picker` rather than a segmented one: a driver may hold several vehicles and the row has
    /// the width of a value, not of a control per option.
    private var vehicleRow: some View {
        GroupedRow(
            titleKey: "tracker_vehicle_label",
            subtitleKey: model.state.hasNoVehicle ? "tracker_no_vehicles" : nil,
            showsSeparator: true
        ) {
            Picker(
                selection: Binding(
                    get: { model.state.selectedVehicleId ?? "" },
                    set: { model.select(vehicleId: $0) }
                )
            ) {
                ForEach(model.state.vehicles, id: \.vehicleId) { vehicle in
                    Text(vehicleLabel(vehicle)).tag(vehicle.vehicleId)
                }
            } label: {
                Text(key: "tracker_vehicle_label")
            }
            .pickerStyle(.menu)
            .labelsHidden()
            .disabled(model.state.vehicles.isEmpty)
        }
    }

    /// The wireframe's `IMEI   8612 3456 …`, as the field D2' §SCR-DI-027's SwiftUI column names.
    private var imeiRow: some View {
        GroupedRow(
            titleKey: "tracker_imei_label",
            subtitleKey: model.state.isImeiRejected ? "tracker_imei_invalid" : "tracker_imei_help",
            showsSeparator: false
        ) {
            TextField(
                "",
                text: Binding(get: { model.state.imei }, set: model.onImeiChange)
            )
            .mageFont(.body)
            .foregroundStyle(model.state.isImeiRejected ? MageRideColor.error : MageRideColor.onSurface)
            .keyboardType(.numberPad)
            .multilineTextAlignment(.trailing)
            .disabled(model.state.isPaired)
            .accessibilityLabel(Text(key: "tracker_imei_label"))
        }
    }

    // MARK: - The button row

    /// The wireframe's `[▣ Scan device QR] [Bind code]`.
    ///
    /// **Bind code** is disabled and the line underneath says why: the owner-facing wrapper carries no
    /// `bindCode`, so there is nowhere for one to be sent. Scan is disabled on a handset whose scanner
    /// cannot run — every simulator, and any device older than the A12 `DataScannerViewController`
    /// needs — rather than hidden, for the same reason.
    private var entryMethods: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
            HStack(spacing: MageRideSpacing.xs) {
                OutlinedAction(
                    labelKey: "tracker_scan_action",
                    symbolName: "qrcode.viewfinder",
                    isEnabled: !model.state.isPaired && model.isScanSupported
                ) {
                    Task { await model.startScan() }
                }

                OutlinedAction(labelKey: "tracker_bind_code_action", isEnabled: false) {}
            }

            Text(key: "tracker_bind_code_help")
                .mageFont(.label)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        }
    }

    // MARK: - The cards

    /// D2' §SCR-DI-027's *"paired → cert-issued confirmation"*.
    ///
    /// The `201` from the bind **is** the credential: provisioning-svc mints the X.509 or the signed
    /// PSK inside that call (T-02, D6' §4.2), so a `bindingId` in hand means one was issued. What this
    /// card cannot show is the device's live health — `lastSeen`, `battery`, `signal` come from
    /// `GET /v1/trackers/{imei}`, which has no client here (C074 spec gap 3).
    private func pairedCard(_ binding: TrackerBinding) -> some View {
        NoticeCard(
            titleKey: "tracker_paired_title",
            symbolName: "antenna.radiowaves.left.and.right",
            accent: MageRideColor.success
        ) {
            Text("tracker_paired_body".localisedFormat(TrackerImei.grouped(binding.imei)))
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        }
    }

    /// The wireframe's green *"✦ Hardware tracker behaviour"* card — and the C092 fence, in words.
    ///
    /// US-3.6: once the device is bound it is the single active publisher and the phone stops.
    /// US-3.22/3.23: a Mode A/B tracker starts and ends journeys on ignition with no app involved, so
    /// the dashboard is *already* at *"Journey started"* when the driver opens it. US-3.21: a Mode C
    /// tracker's GPS is accepted only while the driver is Online.
    private var hardwareBehaviour: some View {
        NoticeCard(
            titleKey: "tracker_behaviour_title",
            symbolName: "sparkles",
            accent: MageRideColor.success
        ) {
            VStack(alignment: .leading, spacing: MageRideSpacing.xxs) {
                ForEach(Self.behaviourKeys, id: \.self) { key in
                    Text(key: key)
                        .mageFont(.bodySmall)
                        .foregroundStyle(MageRideColor.onSurfaceVariant)
                        .frame(maxWidth: .infinity, alignment: .leading)
                }
            }
        }
    }

    /// The wireframe's *"Fleet (5,000+)? Use Admin Portal CSV ›"* (US-3.2, T-09).
    ///
    /// Drawn as a note rather than as a link: the bulk path is `POST /v1/fleets/{fleetId}/trackers/bulk`
    /// on provisioning-svc and it is the **Admin Portal's**, and this app has no Admin Portal origin to
    /// open — ``DriverEnvironment`` carries the gateway and the MQTT host, and D7' puts the portal on a
    /// different one. A `›` that opened nothing would be worse than a sentence that says where to go.
    private var fleetCsv: some View {
        NoticeCard(symbolName: "square.and.arrow.up.on.square", accent: MageRideColor.secondary) {
            Text(key: "tracker_fleet_csv")
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideColor.onSurfaceVariant)
        }
    }

    /// *"Three-wheeler · ABC-1234"* — the type is trilingual and the plate is not.
    private func vehicleLabel(_ vehicle: VehicleSummary) -> String {
        vehicle.vehicleType.labelKey.localised + MageRideSymbols.separator + vehicle.registrationNumber
    }

    private static let behaviourKeys = [
        "tracker_behaviour_publisher",
        "tracker_behaviour_ignition",
        "tracker_behaviour_mode_c",
    ]
}
